### PR TITLE
adding support for temporal subsets

### PR DESCRIPTION
**Pull request by bmorris3**
Downstream, LCviz will show light curves as profiles of flux over time. In order give LCviz a subset API that's consistent with jdaviz, we need to generalize a few things here.

The `get_data_from_viewer` and `get_subsets` methods in `Application` make some assumptions about the dimensionality of data in jdaviz (data with ndim=1 is spectral; dim options are only spectral or spatial). This PR generalizes the `get_data_from_viewer` method to work with non-spectral profiles, and introduces a temporal subset option.  

I've milestoned this to 3.4.1 just in case we can get a release ASAP.

### Change log entry

- [X] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [X] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [X] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [X] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [X] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)? [🐱](https://jira.stsci.edu/browse/JDAT-3297)

### CODE REVIEW

File: jdaviz/app.py [hunk 1]
```diff
@@ -686,8 +687,13 @@ def get_data_from_viewer(self, viewer_reference, data_label=None,
...
                             try:
-                                layer_data = handler.to_object(layer_data,
-                                                               statistic=statistic)
+                                if cls == Spectrum1D:
+                                    # if this is a spectrum, apply the `statistic`:
+                                    layer_data = handler.to_object(layer_data,
+                                                                   statistic=statistic)
+                                else:
+                                    # otherwise simply translate to an object:
+                                    layer_data = handler.to_object(layer_data)
```
[the remark this answers is not in the document]
Comment (bmorris3) [reply jdaviz/app.py:696]: What is a 2D temporal slice?

[the remark this answers is not in the document]
Comment (pllim) [reply jdaviz/app.py:696]: I am guessing like spectral cube but in time domain?

[the remark this answers is not in the document]
Comment (kecnry) [reply jdaviz/app.py:696]: We will eventually want to support target pixel files and modifying the aperture, but if its not obvious how to handle that now, we can just add a note and modify this logic later when needed (I don't think we'll be adding support for that in the terribly near future).

[the remark this answers is not in the document]
Comment (pllim) [reply jdaviz/app.py:696]: Well, I know that is a thing in heliophysics, so definitely possible.

File: jdaviz/app.py [hunk 2]
```diff
@@ -686,8 +687,13 @@ def get_data_from_viewer(self, viewer_reference, data_label=None,
                         if cls is not None:
                             handler, _ = data_translator.get_handler_for(cls)
                             try:
-                                layer_data = handler.to_object(layer_data,
-                                                               statistic=statistic)
+                                if cls == Spectrum1D:
```
Comment (kecnry) [inline jdaviz/app.py:690]: is there any other `cls` that we need to support for collapsed cubes in cubeviz?

**Review by kecnry — APPROVED** [review]
looks reasonable to me, but I'm curious if @javerbukh has any thoughts/concerns as to how this fits into the broader subsets refactoring work.  If we need to merge to move forward with the downstream PR, we can always come back and tweak the implementation here though.

File: jdaviz/app.py (hunk 2 above)
Comment (bmorris3) [reply jdaviz/app.py:690]: There's been talk of visualizing ramps in a Cubeviz-like helper (cc @camipacifici), but I don't think we need to worry too much about supporting that "correctly" yet?

Comment (kecnry) [reply jdaviz/app.py:690]: ok, but right now there is no other class that we support mapping (ndarray, ccddata, etc)?

Comment (bmorris3) [reply jdaviz/app.py:690]: We often check against `cls` is `Spectrum1D` to check if a `Data` object is a spectrum. I think we'd have to do some work to generalize `app.py` to representing ND spectra with other classes.

**Comment by bmorris3** [comment]
I'll check in with @javerbukh about this one at tag up today if I can, and merge if not.

**Comment by bmorris3** [comment]
I just remembered Jesse is out for a while, so I'll merge now. Thanks all!

**Review by rosteen — APPROVED** [review]

**Comment by kecnry** [comment]
@meeseeksdev backport to v3.4.x

**Comment by pllim** [comment]
Was this backported? If so, please cross link the backport PR here. If not, please change the milestone. Thanks!

**Pull request merged**